File: queues/ibmmq/PHP/lib/IbmmqRest.php
<?php
declare(strict_types=1);
/*
 * Copyright Core DF — Apache License 2.0
 */
final class IbmmqRest
{
    private static function baseUrl(): string
    {
        $explicit = getenv('MQ_REST_BASE_URL') ?: '';
        if ($explicit !== '') {
            return rtrim($explicit, '/');
        }
        $host = getenv('MQ_HOST') ?: '';
        $port = getenv('MQ_REST_PORT') ?: '9443';
        return "https://{$host}:{$port}/ibmmq/rest/v2";
    }

    private static function authHeader(): array
    {
        $user = getenv('MQ_USER') ?: '';
        $password = getenv('MQ_PASSWORD') ?: '';
        if ($user === '') {
            return [];
        }
        return ['Authorization: Basic ' . base64_encode("{$user}:{$password}")];
    }

    public static function putMessage(string $queue, mixed $body): array
    {
        $qmgr = getenv('MQ_QUEUE_MANAGER') ?: '';
        $payload = is_string($body) ? $body : json_encode($body);
        $url = self::baseUrl() . "/messaging/qmgr/{$qmgr}/queue/{$queue}/message";
        $ch = curl_init($url);
        curl_setopt_array($ch, [
            CURLOPT_POST => true,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_SSL_VERIFYPEER => false,
            CURLOPT_HTTPHEADER => array_merge(self::authHeader(), ['Content-Type: application/json']),
            CURLOPT_POSTFIELDS => json_encode(['type' => 'string', 'content' => $payload]),
        ]);
        $resp = curl_exec($ch);
        $code = (int)curl_getinfo($ch, CURLINFO_HTTP_CODE);
        curl_close($ch);
        if ($code >= 400) {
            return ['status_code' => $code, 'error' => $resp ?: 'put failed'];
        }
        return ['status_code' => 200];
    }

    public static function getMessages(string $queue, int $maxMessages): array
    {
        $qmgr = getenv('MQ_QUEUE_MANAGER') ?: '';
        $messages = [];
        for ($i = 0; $i < max(1, $maxMessages); $i++) {
            $url = self::baseUrl() . "/messaging/qmgr/{$qmgr}/queue/{$queue}/message";
            $ch = curl_init($url);
            curl_setopt_array($ch, [
                CURLOPT_CUSTOMREQUEST => 'DELETE',
                CURLOPT_RETURNTRANSFER => true,
                CURLOPT_SSL_VERIFYPEER => false,
                CURLOPT_HTTPHEADER => self::authHeader(),
            ]);
            $resp = curl_exec($ch);
            $code = (int)curl_getinfo($ch, CURLINFO_HTTP_CODE);
            curl_close($ch);
            if ($code === 204 || $resp === false || $resp === '') {
                break;
            }
            if ($code >= 400) {
                return ['status_code' => $code, 'error' => $resp];
            }
            $parsed = json_decode($resp, true);
            $value = is_array($parsed) ? ($parsed['content'] ?? $parsed) : $resp;
            if (is_string($value)) {
                $decoded = json_decode($value, true);
                if ($decoded !== null) {
                    $value = $decoded;
                }
            }
            $messages[] = ['queue' => $queue, 'value' => $value];
        }
        return ['status_code' => 200, 'messages' => $messages];
    }
}

File: queues/ibmmq/Ruby/lib/ibmmq_rest.rb
# Copyright Core DF — Apache License 2.0
# IBM MQ REST API helpers (MQ 9.1+ with REST enabled).

require 'json'
require 'net/http'
require 'uri'
require 'base64'
require 'openssl'

module IbmmqRest
  module_function

  def base_url
    explicit = ENV['MQ_REST_BASE_URL']
    return explicit.chomp('/') if explicit && !explicit.empty?

    host = ENV['MQ_HOST']
    port = ENV['MQ_REST_PORT'] || '9443'
    "https://#{host}:#{port}/ibmmq/rest/v2"
  end

  def auth_header
    user = ENV['MQ_USER']
    password = ENV['MQ_PASSWORD']
    return {} if user.to_s.empty?
    token = Base64.strict_encode64("#{user}:#{password}")
    { 'Authorization' => "Basic #{token}" }
  end

  def put_message(queue, body)
    qmgr = ENV['MQ_QUEUE_MANAGER']
    url = URI("#{base_url}/messaging/qmgr/#{qmgr}/queue/#{queue}/message")
    req = Net::HTTP::Post.new(url)
    auth_header.each { |k, v| req[k] = v }
    req['Content-Type'] = 'application/json'
    req.body = JSON.generate(type: 'string', content: body.is_a?(String) ? body : JSON.generate(body))
    resp = Net::HTTP.start(url.host, url.port, use_ssl: url.scheme == 'https', verify_mode: OpenSSL::SSL::VERIFY_NONE) do |h|
      h.request(req)
    end
    return { status_code: resp.code.to_i, error: resp.body } if resp.code.to_i >= 400
    { status_code: 200 }
  end

  def get_messages(queue, max_messages)
    qmgr = ENV['MQ_QUEUE_MANAGER']
    messages = []
    max_messages.times do
      url = URI("#{base_url}/messaging/qmgr/#{qmgr}/queue/#{queue}/message")
      req = Net::HTTP::Delete.new(url)
      auth_header.each { |k, v| req[k] = v }
      resp = Net::HTTP.start(url.host, url.port, use_ssl: url.scheme == 'https', verify_mode: OpenSSL::SSL::VERIFY_NONE) do |h|
        h.request(req)
      end
      break if resp.code.to_i == 204 || resp.body.to_s.empty?
      return { status_code: resp.code.to_i, error: resp.body } if resp.code.to_i >= 400
      parsed = begin; JSON.parse(resp.body); rescue; resp.body; end
      value = parsed.is_a?(Hash) ? (parsed['content'] || parsed) : parsed
      value = begin; JSON.parse(value); rescue; value; end if value.is_a?(String)
      messages << { queue: queue, value: value }
    end
    { status_code: 200, messages: messages }
  end
end
